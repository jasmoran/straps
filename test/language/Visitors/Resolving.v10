#!Resolving.out
#-

include "src/stdlib/stdlib.v10"
include "src/language/parser/Parser.v10"
include "src/language/Scope.v10"
include "src/language/Type.v10"

include "src/language/visitors/Visitor.v10"
include "src/language/visitors/Printer.v10"
include "src/language/visitors/Constructor.v10"
include "src/language/visitors/Typing.v10"
include "src/language/visitors/Method.v10"
include "src/language/visitors/Definition.v10"
include "src/language/visitors/Resolving.v10"

global(Scope scope, Node root_node)

func void main()(Stream input, Parser parser, Node root, Visitor v, Scope sc) {
	# Open input file
	input = File_open("test/language/Visitors/sample.v10.test")
	if (input < 0) {
		"Couldn't read input file".println(STDERR)
		return 1
	}

	parser = Parser_new("sample.v10", input)
	root = parser.parse()
	root_node = root
	sc = Scope_new()
	scope = sc

	v = Constructor_new()
	root.accept(v)
	v = Typing_new()
	root.accept(v)
	v = Method_new()
	root.accept(v)
	v = Definition_new()
	root.accept(v)
	v = Resolving_new()
	root.accept(v)
	v = Printer_new()
	root.accept(v)

	sc.print()
}
