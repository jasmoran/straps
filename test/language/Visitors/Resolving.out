nBLOCK
  nENUM(Boolean)
    nBLOCK
      nNUMBER(<Boolean>: "0")
      nNUMBER(<Boolean>: "1")
  nFUNCTION(<Any>: allocate)
size: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
    nLIST
      nIDENTIFIER(<Number>: "size")
    nLIST
    nBLOCK
  nFUNCTION(<Boolean>: thisisafunction)
arg1: <<
  nBINARYOP(<Any>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
arg2: <<
  nBINARYOP(<Any>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "24")
>>
vrbl1: <<
  nBINARYOP(<Any>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
vrbl2: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-16")
>>
    nLIST
      nIDENTIFIER(<Any>: "arg1")
      nIDENTIFIER(<Any>: "arg2")
    nLIST
      nIDENTIFIER(<Any>: "vrbl1")
      nIDENTIFIER(<Number>: "vrbl2")
    nBLOCK
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Any>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "16")
        nBINARYOP(<Any>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
      nUNARYOP(Unknown: INCREMENT, right)
        nBINARYOP(<Number>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-16")
      nBINARYOP(Unknown: ADD)
        nBINARYOP(<Any>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
        nNUMBER(<Number>: "1337")
      nBINARYOP(<Any>: DOT)
        nBINARYOP(<Any>: PAREN_OPEN)
          nFUNCTION(<Any>: _String_println)
this: <<
  nBINARYOP(<String>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
stream: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "24")
>>
ilikenumbers: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "32")
>>
            nLIST
              nIDENTIFIER(<String>: "this")
              nIDENTIFIER(<Number>: "stream")
              nIDENTIFIER(<Number>: "ilikenumbers")
            nLIST
            nBLOCK
          nCALL
            nBINARYOP(<String>: PAREN_OPEN)
              nFUNCTION(<String>: _Character_toString)
this: <<
  nBINARYOP(<Character>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
                nLIST
                  nIDENTIFIER(<Character>: "this")
                nLIST
                nBLOCK
              nCALL
                nBINARYOP(<Character>: DOT)
                  nBINARYOP(<String>: DOT)
                    nNUMBER(<Boolean>: "1")
                    nINDEX(<String>: "0")
                  nINDEX(<Character>: "0")
            nNUMBER(<Number>: "77")
            nBINARYOP(<Number>: SQUARE_OPEN)
              nREGISTER(Unknown: "rbp")
              nNUMBER(#Number#: "-16")
        nINDEX(<Any>: "0")
      nLOOP
        nBLOCK
          nBINARYOP(<Any>: PAREN_OPEN)
            nFUNCTION(<Any>: _String_println)
this: <<
  nBINARYOP(<String>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
stream: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "24")
>>
ilikenumbers: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "32")
>>
              nLIST
                nIDENTIFIER(<String>: "this")
                nIDENTIFIER(<Number>: "stream")
                nIDENTIFIER(<Number>: "ilikenumbers")
              nLIST
              nBLOCK
            nCALL
              nSTRING(<String>: "oopsy daisy")
              nNUMBER(<Number>: "1")
              nNUMBER(<Number>: "2")
      nRETURN
        nNUMBER(<Boolean>: "1")
  nFUNCTION(<Any>: Any_new)
deref: <<
  nBINARYOP(<Any>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
obj: <<
  nBINARYOP(<Any>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
      nIDENTIFIER(<Any>: "deref")
    nLIST
      nIDENTIFIER(<Any>: "obj")
    nBLOCK
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Any>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
        nBINARYOP(<Any>: PAREN_OPEN)
          nFUNCTION(<Any>: allocate)
size: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
            nLIST
              nIDENTIFIER(<Number>: "size")
            nLIST
            nBLOCK
          nCALL
            nNUMBER(<Number>: "8")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Any>: DOT)
          nBINARYOP(<Any>: SQUARE_OPEN)
            nREGISTER(Unknown: "rbp")
            nNUMBER(#Number#: "-8")
          nINDEX(<Any>: "0")
        nBINARYOP(<Any>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "16")
      nRETURN
        nBINARYOP(<Any>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
  nFUNCTION(<Character>: Character_new)
obj: <<
  nBINARYOP(<Character>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
      nIDENTIFIER(<Character>: "obj")
    nBLOCK
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Character>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
        nBINARYOP(<Any>: PAREN_OPEN)
          nFUNCTION(<Any>: allocate)
size: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
            nLIST
              nIDENTIFIER(<Number>: "size")
            nLIST
            nBLOCK
          nCALL
            nNUMBER(<Number>: "0")
      nRETURN
        nBINARYOP(<Character>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
  nFUNCTION(<Number>: Number_new)
obj: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
      nIDENTIFIER(<Number>: "obj")
    nBLOCK
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Number>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
        nBINARYOP(<Any>: PAREN_OPEN)
          nFUNCTION(<Any>: allocate)
size: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
            nLIST
              nIDENTIFIER(<Number>: "size")
            nLIST
            nBLOCK
          nCALL
            nNUMBER(<Number>: "0")
      nRETURN
        nBINARYOP(<Number>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
  nFUNCTION(<Boolean>: Boolean_new)
as_string: <<
  nBINARYOP(<String>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
obj: <<
  nBINARYOP(<Boolean>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
      nIDENTIFIER(<String>: "as_string")
    nLIST
      nIDENTIFIER(<Boolean>: "obj")
    nBLOCK
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Boolean>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
        nBINARYOP(<Any>: PAREN_OPEN)
          nFUNCTION(<Any>: allocate)
size: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
            nLIST
              nIDENTIFIER(<Number>: "size")
            nLIST
            nBLOCK
          nCALL
            nNUMBER(<Number>: "8")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<String>: DOT)
          nBINARYOP(<Boolean>: SQUARE_OPEN)
            nREGISTER(Unknown: "rbp")
            nNUMBER(#Number#: "-8")
          nINDEX(<String>: "0")
        nBINARYOP(<String>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "16")
      nRETURN
        nBINARYOP(<Boolean>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
  nFUNCTION(<String>: String_new)
random_char: <<
  nBINARYOP(<Character>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
obj: <<
  nBINARYOP(<String>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
      nIDENTIFIER(<Character>: "random_char")
    nLIST
      nIDENTIFIER(<String>: "obj")
    nBLOCK
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<String>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
        nBINARYOP(<Any>: PAREN_OPEN)
          nFUNCTION(<Any>: allocate)
size: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
            nLIST
              nIDENTIFIER(<Number>: "size")
            nLIST
            nBLOCK
          nCALL
            nNUMBER(<Number>: "8")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Character>: DOT)
          nBINARYOP(<String>: SQUARE_OPEN)
            nREGISTER(Unknown: "rbp")
            nNUMBER(#Number#: "-8")
          nINDEX(<Character>: "0")
        nBINARYOP(<Character>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "16")
      nRETURN
        nBINARYOP(<String>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
  nFUNCTION(<String>: _Character_toString)
this: <<
  nBINARYOP(<Character>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
    nLIST
      nIDENTIFIER(<Character>: "this")
    nLIST
    nBLOCK
  nFUNCTION(<Any>: _String_println)
this: <<
  nBINARYOP(<String>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
stream: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "24")
>>
ilikenumbers: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "32")
>>
    nLIST
      nIDENTIFIER(<String>: "this")
      nIDENTIFIER(<Number>: "stream")
      nIDENTIFIER(<Number>: "ilikenumbers")
    nLIST
    nBLOCK
Any: <<
  nCLASS(Any)
deref: <<
  nINDEX(<Any>: "0")
>>
    nLIST
    nBLOCK
>>
Character: <<
  nCLASS(Character)
toString: <<
  nFUNCTION(<String>: _Character_toString)
this: <<
  nBINARYOP(<Character>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
    nLIST
    nLIST
    nBLOCK
>>
    nLIST
    nBLOCK
>>
Number: <<
  nCLASS(Number)
    nLIST
    nBLOCK
>>
Boolean: <<
  nCLASS(Boolean)
as_string: <<
  nINDEX(<String>: "0")
>>
    nLIST
    nBLOCK
>>
String: <<
  nCLASS(String)
println: <<
  nFUNCTION(<Any>: _String_println)
this: <<
  nBINARYOP(<String>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
stream: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "24")
>>
ilikenumbers: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "32")
>>
    nLIST
    nLIST
    nBLOCK
>>
random_char: <<
  nINDEX(<Character>: "0")
>>
    nLIST
    nBLOCK
>>
false: <<
  nNUMBER(<Boolean>: "0")
>>
true: <<
  nNUMBER(<Boolean>: "1")
>>
allocate: <<
  nFUNCTION(<Any>: allocate)
size: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
    nLIST
    nLIST
    nBLOCK
>>
thisisafunction: <<
  nFUNCTION(<Boolean>: thisisafunction)
arg1: <<
  nBINARYOP(<Any>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
arg2: <<
  nBINARYOP(<Any>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "24")
>>
vrbl1: <<
  nBINARYOP(<Any>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
vrbl2: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-16")
>>
    nLIST
    nLIST
    nBLOCK
>>
Any_new: <<
  nFUNCTION(<Any>: Any_new)
deref: <<
  nBINARYOP(<Any>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
obj: <<
  nBINARYOP(<Any>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
    nBLOCK
>>
Character_new: <<
  nFUNCTION(<Character>: Character_new)
obj: <<
  nBINARYOP(<Character>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
    nBLOCK
>>
Number_new: <<
  nFUNCTION(<Number>: Number_new)
obj: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
    nBLOCK
>>
Boolean_new: <<
  nFUNCTION(<Boolean>: Boolean_new)
as_string: <<
  nBINARYOP(<String>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
obj: <<
  nBINARYOP(<Boolean>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
    nBLOCK
>>
String_new: <<
  nFUNCTION(<String>: String_new)
random_char: <<
  nBINARYOP(<Character>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
obj: <<
  nBINARYOP(<String>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
    nBLOCK
>>
_Character_toString: <<
  nFUNCTION(<String>: _Character_toString)
this: <<
  nBINARYOP(<Character>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
    nLIST
    nLIST
    nBLOCK
>>
_String_println: <<
  nFUNCTION(<Any>: _String_println)
this: <<
  nBINARYOP(<String>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
stream: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "24")
>>
ilikenumbers: <<
  nBINARYOP(<Number>: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "32")
>>
    nLIST
    nLIST
    nBLOCK
>>
